caddyhttp/fastcgi/fastcgi.go

```diff
@@ -91,14 +124,18 @@ func (h Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error)
 				resp, err = fcgiBackend.Post(env, r.Method, r.Header.Get("Content-Type"), r.Body, contentLength)
 			}
 
-			if resp.Body != nil {
-				defer resp.Body.Close()
-			}
-
 			if err != nil && err != io.EOF {
+				persistentConnections[network+address].Close()
+				persistentConnections[network+address] = nil
```

Comment:
The garbage collector cleans up memory that is no longer reachable; map entries are very definitely still accessible.

The locks you've added appear to lock the FCGIClient for a specific backend, but these map reads/writes happen unsynchronized for each request that comes in. If one of them does happen concurrently with another, the runtime will panic. (It won't kill your server, because Caddy has a recover higher up, but it will kill the request.)

Consider a RWMutex to protect the map during concurrent access too.